libgadget/init.c

```diff
@@ -315,4 +318,6 @@ setup_smoothinglengths(int RestartSnapNum)
 #ifdef DENSITY_INDEPENDENT_SPH
     density_update();
 #endif //DENSITY_INDEPENDENT_SPH
+
+    if(force_tree_allocated()) force_tree_free();
```

Comment:
You wanted to limit the life cycle of tree inside this function?
force_tree_allocated should be true here.

Reply:
It should, I was just being defensive for future double-frees. I can make that line force_tree_free() if you like. 

The tree needs to be rebuilt after the first domain exchange in run.c anyway (although the first domain will exchange zero particles, so this isn't strictly necessary). Limiting the tree lifetime to this function seemed to minimize special-casing.